adr/ADR-50.md

```diff
@@ -91,7 +90,82 @@ The `LastMsgId` header is currently not supported. A batch will be rejected if t
+
+ * The client will set up a Inbox subscription that will be used for the duration of the batch
+ * A batch will be started by adding the `Nats-Fast-Batch-Id:uuid`, `Nats-Flow:10` and `Nats-Batch-Sequence:1` headers using a *request* with reply being the above Inbox, the server will reply with error or zero byte message. Maximum length of the ID is 64 characters.
+ * Following messages in the same batch will include the `Nats-Fast-Batch-Id:uuid` header and increment `Nats-Batch-Sequence:n` by one, and might optionally include a reply set to the same Inbox to force a immediate zero byte reply
+ * If the final message has the headers `Nats-Batch-Id:uuid`, `Nats-Batch-Sequence:n` and `Nats-Batch-Commit:1`, the server will store the message, commit the batch and reply with a pub ack.
+ * Otherwise, the final message will have headers `Nats-Batch-Id:uuid`, `Nats-Batch-Sequence:n` and `Nats-Batch-Commit:eob` and the server will commit the batch without storing the message and reply with a pub ack.
+ * Clients will monitor the acks and should an ack have a `Nats-Flow` header that differs from the active one they will adjust accordingly
+
+The `Nats-Flow` message has a few formats see the later section.
+
+The server will acknowledge in the following manner:
+
+ * The initial message will get an error - for example, feature not supported - or zero byte ack
+ * The server will then send zero byte acks back based on the values of `Nats-Flow`, the `Nats-Flow` header will always be included, such acks can optionally include a header `Nats-Batch-Sequence:n` which indicates which message triggered the ack
+ * The final message will get a pub ack as described later
+
+By always sending the current flow header back we guard against lost acks.
+
+### Server Errors
+
+The server will respond with the following errors if committing a batch fails:
+
+| ErrCode | Code | Description                                                         |
+|---------|------|---------------------------------------------------------------------|
+| 10174   | 400  | Batch publish not enabled on stream                                 |
+| 10176   | 400  | Batch publish is incomplete and was abandoned                       |
+| 10179   | 400  | Batch publish ID is invalid (exceeds 64 characters)                 |
+| 10175   | 400  | Batch publish sequence is missing                                   |
+| 10177   | 400  | Batch publish unsupported header used (`Nats-Expected-Last-Msg-Id`) |
+| 10201   | 400  | Batch publish contains duplicate message id (`Nats-Msg-Id`)         |
+
+### Server Behavior Design
+
+* The server will limit the `Nats-Fast-Batch-Id` to 64 characters and response with a error Pub Ack if its too long
+* Server will reject messages for which the batch is unknown with a error Pub Ack
+* If messages in a batch is received and any gap is detected an ack will be sent back indicating the gap
```

Comment:
In which case - thw whole batch should be aborted and restarted on client, while server drops the whole batch? 
Does the error contain range of gaps?

Reply:
These kinds of batches are not staged or reliable, not atomically committed, so could abandon or could communicate the missed range - we suspect in this use case its like a stream of metrics, you aren't going to care. But for sure one to discuss